Engine/ac/dynobj/managedobjectpool.cpp

```diff
@@ -65,166 +47,155 @@ int32_t ManagedObjectPool::AddRef(int32_t handle) {
 {
-    ManagedObjectLog("Running garbage collection");
-
-    for (int i = 1; i < numObjects; i++) 
+    for (auto it = objects.begin(); it != objects.end() /* not hoisted */; /* no increment */)
     {
-        if ((objects[i].refCount < 1) && (objects[i].callback != NULL)) 
-        {
-            Remove(i);
+        auto & o = it->second;
+        it++; // iterator is invalid _after_ erase, so increment now.
+        if (o.refCount < 1) {
+            Remove(o.handle);
         }
     }
+    ManagedObjectLog("Ran garbage collection");
 }
 
-int ManagedObjectPool::AddObject(const char *address, ICCDynamicObject *callback, bool plugin_object, int useSlot) {
-    if (useSlot == -1)
-        useSlot = numObjects;
+inline int handle_increment(int handle) {
+    if (handle >= INT_MAX || handle <= 0) {
+        return 1;
+    }
+    return handle + 1;
+}
 
-    objectCreationCounter++;
+int ManagedObjectPool::AddObject(const char *address, ICCDynamicObject *callback, bool plugin_object, int useSlot) 
+{
+    int handle;
 
-    if (useSlot < arrayAllocLimit) {
-        // still space in the array, so use it
-        Init(useSlot, address, callback, plugin_object ? kScValPluginObject : kScValDynamicObject);
-        if (useSlot == numObjects)
-            numObjects++;
-        return useSlot;
-    }
-    else {
-        // array has been used up
-        if (useSlot == numObjects) {
-            // if adding new (not un-serializing) check for empty slot
-            // check backwards, since newer objects don't tend to last
-            // long
-            for (int i = arrayAllocLimit - 1; i >= 1; i--) {
-                if (objects[i].handle == 0) {
-                    Init(i, address, callback, plugin_object ? kScValPluginObject : kScValDynamicObject);
-                    return i;
-                }
-            }
+    if (useSlot >= 0) {
+        if (objects.find(useSlot) != objects.end()) {
+            cc_error("Slot used: %d", useSlot);
+            return -1;
+        }
+        handle = useSlot;
+    } else {
+        handle = nextHandle;
+        while (objects.find(handle) != objects.end()) {
+            handle = handle_increment(handle);
         }
-        // no empty slots, expand array
-        while (useSlot >= arrayAllocLimit)
-            arrayAllocLimit += ARRAY_INCREMENT_SIZE;
-
-        objects = (ManagedObject*)realloc(objects, sizeof(ManagedObject) * arrayAllocLimit);
-        memset(&objects[useSlot], 0, sizeof(ManagedObject) * ARRAY_INCREMENT_SIZE);
-        Init(useSlot, address, callback, plugin_object ? kScValPluginObject : kScValDynamicObject);
-        if (useSlot == numObjects)
-            numObjects++;
-        return useSlot;
     }
+
+    objects.insert({handle, {
+        ManagedObject {
+            obj_type: plugin_object ? kScValPluginObject : kScValDynamicObject,
+            handle: handle,
+            addr: address,
+            callback: callback,
+            refCount: 0
+        }
+    }});
```

Comment:
This does not compile using MSVS 2015 compiler. What version of standard is this kind of initialization?

Reply:
Interesting. I'm using MSVS 2017.  It's possible I didn't test this under Windows and I've inadvertently used a gcc extension. I was aiming for this: https://docs.microsoft.com/en-us/cpp/cpp/initializing-classes-and-structs-without-constructors-cpp?view=vs-2017

Reply:
Pushed a fix.